search/graph_search.go

```diff
@@ -382,59 +381,85 @@ func CopyDirectedGraph(dst graph.MutableDirectedGraph, src graph.DirectedGraph)
 // An undirected graph should end up with as many SCCs as there are "islands" (or subgraphs) of
 // connections, meaning having more than one strongly connected component implies that your graph
 // is not fully connected.
-func Tarjan(g graph.Graph) (sccs [][]graph.Node) {
-	index := 0
-	stackSet := make(intSet)
-	vStack := &nodeStack{}
-	sccs = make([][]graph.Node, 0)
-
+func Tarjan(g graph.Graph) [][]graph.Node {
 	nodes := g.NodeList()
-	lowlinks := make(map[int]int, len(nodes))
-	indices := make(map[int]int, len(nodes))
+	t := tarjan{
+		succ: setupFuncs(g, nil, nil).successors,
 
-	successors := setupFuncs(g, nil, nil).successors
+		indexTable: make(map[int]int, len(nodes)),
+		lowLink:    make(map[int]int, len(nodes)),
+		isRemoved:  make(intSet, len(nodes)),
+	}
+	for _, v := range nodes {
+		if t.indexTable[v.ID()] == 0 {
+			t.strongconnect(v)
+		}
+	}
+	return t.sccs
+}
 
-	var strongconnect func(graph.Node) []graph.Node
+// tarjan implements Tarjan's strongly connected component finding
+// algorithm. The implementation is from the pseudocode at
+//
+// http://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
```

Comment:
It would be good to add a date accessed as part of the reference here.  Wikipedia changes.